### PR TITLE
Add api layer and special files

### PR DESCRIPTION
**Pull request by libre-man**
Add an api socket and special files.

### CODE REVIEW

File: cgfs.py [hunk 1]
```diff
@@ -102,36 +143,707 @@ def pop(self, filename):
...
+        self.full_path = path.join(self._tmpdir, self._filename)
+        Path(path.join(self.full_path)).touch()
+
+    @property
+    def stat(self):
+        st = os.lstat(self.full_path)
+        stat = {
+            key: getattr(st, key)
+            for key in (
+                'st_atime', 'st_ctime', 'st_gid', 'st_mode', 'st_mtime',
```
Comment (olmokramer) [inline cgfs.py]: `st_mode` can be removed here as it is set later on.

File: cgfs.py [hunk 2]
```diff
@@ -102,36 +143,707 @@ def pop(self, filename):
...
+    @property
+    def data(self):
+        if self._data is None:
+            self._data = cgapi.get_file(self.id)
+            self.stat['st_size'] = len(self._data)
+        return self._data
+
+    @data.setter
+    def data(self, data):
+        self._data = data
```
Comment (olmokramer) [inline cgfs.py]: Should also set `self.stat['st_size']` here.

File: cgfs.py [hunk 3]
```diff
@@ -160,34 +870,205 @@ def truncate(self, length):
...
+        with self._lock:
+            self.files.getattr()
+            self.files.insert(self.special_socketfile)
+            self.files.insert(
+                SpecialFile(
+                    '.cg-mode', b'FIXED\n' if self.fixed else b'NOT_FIXED\n'
+                )
+            )
+            self.load_courses()
+        print('Mounted')
```
Comment (olmokramer) [inline cgfs.py]: @olmokramer wrote:

> This should only be printed in `--verbose` mode in my opinion. I would probably run the filesystem in the background (with `cgfs user mountpoint &`) and it's very annoying when backgrounded programs print to stdout.

@libre-man wrote:

> No this should only not be printed in `--daemon` mode imo.

That would be fine. Couldn't find a `--daemon` mode, though?

File: cgfs.py [hunk 4]
```diff
@@ -315,38 +1227,61 @@ def create(self, path, mode):
...
             if isinstance(file, Directory):
                 query_path += '/'
         else:
             submission = None
             query_path = None
 
-        return file.getattr(submission, query_path)
+        attrs = file.getattr(submission, query_path)
+        if self.fixed and isinstance(file, File):
+            attrs['st_mode'] &= ~0o222
```
Comment (olmokramer) [inline cgfs.py]: Shouldn't we use `0o220` here instead?

File: cgfs.py [hunk 5]
```diff
@@ -365,112 +1304,138 @@ def mkdir(self, path, mode):
...
+        file = self.get_file(parts[-1], start=parent, expect_type=SingleFile)
 
-        if file.data is None:
-            file.open(cgapi.get_file(file.id))
+        if isinstance(file, (TempFile, SpecialFile)):
+            file.open()
 
-        if flags & O_TRUNC:
+        # This is handled by fuse [0] but it can be disabled so it is better to
+        # be save than sorry as it can be enabled.
```
Comment (olmokramer) [inline cgfs.py]: s/save/safe

File: cgfs.py [hunk 6]
```diff
@@ -365,112 +1304,138 @@ def mkdir(self, path, mode):
...
+        self._open_files[self.fd] = file
         return self.fd
 
     def read(self, path, size, offset, fh):
-        file = self.get_file(path, expect_type=File)
-        return file.data[offset:offset + size]
+        with self._lock:
+            file = self._open_files[fh]
+
+            if isinstance(file, (TempFile, SpecialFile)):
```
Comment (olmokramer) [inline cgfs.py]: Maybe we should make a `read` method on the `File` class instead of special-casing the `TempFile`?

File: cgfs.py [hunk 7]
```diff
@@ -491,38 +1456,72 @@ def symlink(self, target, source):
...
+            file = self.get_file(path)
+
+            assert file is not None
 
-        atime, mtime = times or (time(), time())
-        file.setattr('st_atime', atime)
-        file.setattr('st_mtime', mtime)
+            atime, mtime = times or (time(), time())
+
+            if isinstance(file, (TempFile, SpecialFile)):
```
Comment (olmokramer) [inline cgfs.py]: Maybe we should add a `utimens` method on the `File` class instead of having a special case for `TempFile`?

File: cgfs.py [hunk 8]
```diff
@@ -491,38 +1456,72 @@ def symlink(self, target, source):
...
+                raise FuseOSError(EPERM)
+            else:
+                file.setattr('st_atime', atime)
+                file.setattr('st_mtime', mtime)
 
     def write(self, path, data, offset, fh):
-        file = self.get_file(path, expect_type=File)
-        return file.write(data, offset)
+        with self._lock:
+            file = self._open_files[fh]
```
Comment (olmokramer) [inline cgfs.py]: Shouldn't we check here if the file handle is actually open?

File: cgfs.py [hunk 9]
```diff
@@ -542,8 +1541,18 @@ def write(self, path, data, offset, fh):
         metavar='PASSWORD',
         type=str,
         dest='password',
-        help='Your CodeGra.de password, don\' pass this option if you want to '
-        ' pass your password over stdin.'
+        help="""Your CodeGra.de password, don' pass this option if you want to
```
Comment (olmokramer) [inline cgfs.py]: s/don'/don't

File: cgfs.py [hunk 10]
```diff
@@ -561,16 +1570,61 @@ def write(self, path, data, offset, fh):
...
+        help="""Mount the original files as read only. It is still possible to
+        create new files, but it is not possible to alter existing files."""
+    )
+    argparser.add_argument(
+        '-r',
+        '--rubric-edit',
+        dest='rubric_append_only',
+        action='store_false',
+        default=True,
+        help='''Make it possible to delete rubric items or categories using the
```
Comment (olmokramer) [inline cgfs.py]: Why no double quotes?

**Comment by olmokramer** [comment]
Moved over some comments from the old PR #3

File: cgfs.py (hunk 3 above)
Comment (libre-man) [reply cgfs.py]: No that is expected as it is not a valid flag at the moment.

File: cgfs.py (hunk 4 above)
Comment (libre-man) [reply cgfs.py]: Why?

File: cgfs.py (hunk 6 above)
Comment (libre-man) [reply cgfs.py]: Yes.

File: cgfs.py (hunk 8 above)
Comment (libre-man) [reply cgfs.py]: No, as it is removed form the `_open_files` during a release.

File: cgfs.py (hunk 10 above)
Comment (libre-man) [reply cgfs.py]: Because this is wrong.

**Comment by libre-man** [comment]
Thanks!

File: cgfs.py (hunk 4 above)
Comment (olmokramer) [reply cgfs.py]: So we don't have to rely on the underlying filesystem to throw an error when another user tries to write to it

File: cgfs.py (hunk 3 above)
Comment (olmokramer) [reply cgfs.py]: Ok fine

File: api_consumer.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
+#!/usr/bin/env python3
+import os
+import sys
+import json
+import socket
+
+
+def print_usage():
```
Comment (olmokramer) [inline api_consumer.py]: I'd prefer using the `argparse` package, but we can do that in another PR

File: api_consumer.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+def recv(s):
+    message = b''
+
+    while True:
+        m = s.recv(1024)
+        message += m
+        if len(m) < 1024:
+            break
+    return message
+
```
Comment (olmokramer) [inline api_consumer.py]: Missing newline?

File: api_consumer.py [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+                    'file': os.path.abspath(file),
+                }
+            ).encode('utf8')
+        )
+    )
+    if json.loads(recv(s))['ok']:
+        return 0
+    else:
+        return 2
+
```
Comment (olmokramer) [inline api_consumer.py]: Missing newline?

File: cgapi.py [hunk 14]
```diff
@@ -44,105 +120,195 @@ def __init__(self, response):
...
+        self._handle_response_error(r)
+
+        return r.json()
+
+    def select_rubricitems(self, submission_id, items):
+        url = self.routes.select_rubricitems(submission_id)
+        r = self.s.patch(url, json={'items': items})
+        self._handle_response_error(r)
+
+    def get_feedbacks(self, assignment_id):
```
Comment (olmokramer) [inline cgapi.py]: Maybe we should call this `get_line_feedback` or something, to better differentiate it from `get_feedback`? And also because "feedbacks" isn't really a word.

Comment (libre-man) [reply cgapi.py]: WRONG, feedbacks is a word.

Comment (olmokramer) [reply cgapi.py]: Hmm I found some sources staying that it has indeed become acceptable to use feedbacks as a plural form of feedback, so fair enough. But still it sounds a bit awkward and I'd still appreciate some differentiation with `get_feedback`. As it is now I would expect this function to do the equivalent of multiple calls to that function.

Comment (libre-man) [reply cgapi.py]: Oh wow, I was just making a joke. A different name would be fine by me :)

Comment (olmokramer) [reply cgapi.py]: LOL. You were correct though, it seems
